Add AIP 129 for server modified values (#1144)
googleapis/aip: aip/general/0129.md

---
id: 129
state: draft
created: 2023-06-21
placement:
  category: resource-design
  order: 65
---

# Server-Modified Values and Defaults

> Server-modified and default values tend to cause problems for

Reviewer: ```suggestion
Server-modified and default values often result in a poor user experience for
```

I think "problems" is kind of generic, this wording might be a little clearer.

[state-driven clients](state-driven-clients). These clients are often unable to
tell when their desired state matches the current state for these fields, as the

> rules by which a server may modify and return values are not public and

Reviewer: ```suggestion
rules by which a server may modify and return values are complex, not public and not
```

repeatable. This AIP aims to clarify what behaviors can be expected from
services and describing how they can be repeated on the client-side.

## Guidance

### Effective Values

> There are instances where a service will allocate, generate, or calculate a

Reviewer: spacing: please add a newline after the header.

value if the client chooses not to specify one. For example: a client may
create a virtual machine without specifying a static IP address for the virtual
machine to be available on. Such a scenario is opting into dynamic IP address
allocation.

Some examples of these types of fields are ones that are:

* generated (UUID)
* allocated (dynamic IP address)
* assigned (most recent software package version)

An attribute with an effective value ***must*** be expressed as two fields in
the API:
* a mutable field that can be optionally set by the user and **must not** be
modified by the service
* an `OUTPUT_ONLY` field that records the effective value decided on by the
service

Example:
```proto
message VirtualMachine {
  …
  string ip_address = 4;
  string effective_ip_address = 5 [
    (google.api.field_behavior) = OUTPUT_ONLY
  ];
}
```

#### Naming

Effective values *should* be named by prefixing `effective_` to the mutable

> field's name. Consistent naming is important for identifying standard behavior

Reviewer: The "consistent naming" part is a rationale, so it could be moved to a separate rationale section.

That aside, as we already know, a *should* isn't enough for a client to infer standard behavior. 

If it's intended to be programmatically inferable, then it's a *must*. Otherwise I'd drop the statement that says it's used to identify standard behavior.

Author: Moved to rationale & changed to **must**

across APIs and fields.

> ### Client-Settable Fields and Normalization

Reviewer: are we aligning on the term `client-settable`? if so we should refactor AIP-133: https://google.aip.dev/133#user-specified-ids. 

FWIW I think we should just the term `user-specified` since AIP-133 already coined that term, but I do agree that client-settable is a bit clearer.

Also a conversation for @bgrant0607.

Author: `user-specified` sounds better, I'll move to that term

For client-settable fields, the value in response from the service **must** be
the same as provided by the create or update request. For string fields this

> means returning the unchanged character sequence, with two exceptions: 

Reviewer: ```suggestion
means returning the value unchanged, with two exceptions: 
```

nit: "character sequence" is a bit removed from I think what we're saying here.

* When a string field has a data type annotation, a normalized string that
represents the given value may be returned.

> * When a string field has a default value annotation, that default value may be

Reviewer: there is no default value annotation today, correct? 

if not, we cannot recommend that guidance to readers. So we should work on adding that first.

Author: Correct. This should not be merged until that exists.

I can break this into a separate PR not mentioning the new annotations, but my understanding is that this proposal without the `value_type` annotation will not be approved. If I'm off base I can make those changes

Author: Changed wording to talk about the default keyword instead of annotation

returned when the given value is empty or missing.

> Normalizations are important to allow services to store and return values in a

Reviewer: move "normalization" to it's own section? it's a whole topic in it's own right.

Author: Moved to its own section and moved most of this into a rationale section

identical. Normalizing a value on the service side allows the service to accept
a wider range of semantically identical inputs without needing to maintain every
value as a raw string. Surfacing the normalization that is being applied to
clients allows for client side comparison of sent and retrieved values to check
for differences.

For example, in a resource that accepts an email address on a particular field
a client may specify a given email address in a variety of ways. For the email
`ada@example.com` a client may choose to specify `ADA@example.com`,
`aDa@example.com`, or `AdA@example.com`. These are semantically identical and
*should* all be accepted by the service. The service then may choose to
normalize the email address for storage and retrieval through downcasing or
canonicalization. Importantly, the information surfaced to clients on the
normalization of a field must not describe the normalization algorithm itself,
but instead the comparison method used to accurately compute if two values
should be considered equal.

#### Allowed Normalizations

This list of allowed normalizations is defined as follows:

* uuid
* ipv4
* ipv6
* email

> Normalizations on fields **must** be described using the `google.api.value_type`

Reviewer: I don't see `google.api.value_type` declared anywhere - is this something that needs to be added?

Author: Yes, this needs to be introduced.

annotation.

#### Field Value Handling

For all values that do not have an allowed normalization applied

> [state-driven clients](state-driven-clients) will not be able to identify which

Reviewer: we don't need to mention state-driven-clients here. I'd suggest moving a lot of the discussion around state-driven-clients to a `rationale` section, so the guidance is easy to read.

Those who want to understand *why* they need to adhere to this guidance can go to the rationale section.

Author: Done

changes are semantically meaningful. To avoid such mismatching values:

> * service must define and validate a single representation of a field. 

Reviewer: ```suggestion
* service **must** define and validate a single representation of a field. 
```

use RFC-2119 should/must and bold them: https://google.aip.dev/8#requirement-keywords

Reviewer: same goes for other uses of must / should, if they're intended to be RFC-2119-style requirements.

* service must return an error when a client submitted value doesn't match
defined format.

> * service must return the field's value formatted using the same criteria as the

Reviewer: is this last guidance necessary? since the value should be the unchanged value that the user submitted, and that should match the specified format.

It feels redundant.

Author: Fair.

---
id: 129
state: draft
created: 2023-09-20
placement:
  category: resource-design
  order: 65
---

# Server-Modified Values and Defaults

> Services often provide default values for resource fields, and occasionally normalize the user input before returning it in the response. The guidance herein describes how services document such behavior for the benefit of consumers.

Reviewer 2: Nit: Line is too long, please wrap at ~column 80

An attribute with an effective value ***must*** be expressed as two fields in
the API:
* a mutable field that can be optionally set by the user and **must not** be
modified by the service
* an `OUTPUT_ONLY` field that records the effective value decided on by the
service

Example:
```proto
message VirtualMachine {
  …
  string ip_address = 4;
  string effective_ip_address = 5 [
    (google.api.field_behavior) = OUTPUT_ONLY
  ];
}
```

#### Naming

Effective values **must** be named by prefixing `effective_` to the mutable
field's name.

### User-Specified Fields

For user-specified fields, the value in response from the service **must** be
the same as provided by the create or update request. For string fields this
means returning the value unchanged, with two exceptions: 

* When a string field has a data type annotation, a normalized string that
  represents the given value **may** be returned.
* When a string field has a default value indicated by the `default` keyword,
  that default value **may** be returned when the given value is empty or
  missing.

### Normalizations

A field that is normalized by the service **must** be annotated with the

> `google.api.field` extension. The allowed set of normalizations includes

Reviewer 2: Maybe include the following:
`See [AIP-202](./0202.md) for guidance on using this extension`

I think that linking will work, but take a look at other examples for best AIP linking markdown

Author: Added

> `google.api.field` extension. The allowed set of normalizations includes

Reviewer 2: ```suggestion
`google.api.field_info` extension. The allowed set of normalizations includes
```

We had to rename it to avoid `google.protobuf.Field` collisions/confusions.

the following formats:

* uuid
* ipv4
* ipv6
* email

> Normalizations on fields **must** be described using the `google.api.field`

Reviewer 2: ```suggestion
Normalizations on fields **must** be described using the `google.api.field_info`
```

We had to rename it to avoid `google.protobuf.Field` collisions/confusions.

---
id: 129

> state: draft

Reviewer: please update state to approved.

> created: 2023-06-21

Reviewer: if date could be updated too while you're at it, it would be great.

placement:
  category: resource-design
  order: 65
---

# Server-Modified Values and Defaults

Server-modified and default values often result in a poor user experience for
[state-driven clients](state-driven-clients). These clients are often unable to
tell when their desired state matches the current state for these fields, as the
rules by which a server may modify and return values are complex, not public,
and not repeatable. This AIP aims to clarify what behaviors can be expected from

> services and describing how they can be repeated on the client-side.

Reviewer 2: Can this section just give context and not rationale? For example, this section could say something like:
> Services often provide default values for resource fields, and occasionally normalize the user input before returning it in the response. The guidance herein describes how services document such behavior for the benefit of consumers.

Then the [Rationale section](https://google.aip.dev/8#document-structure) discusses the why, for example:
> Server-modified and default values make it harder to implement and use state-driven clients. These clients are...

This is great reasoning, and it is exactly why we added the Rationale section.

Author: Added to base Rationale section, does that look right?

## Guidance

### Single Owner Fields

> Fields should have a single owner, whether that is the client or the server.

Reviewer 2: ```suggestion
Fields **should** have a single owner, whether that is the client or the server.
```

If this is guidance, it **must** be bolded to be a requirement word.

> Server owned fields should be indicated with the `OUTPUT_ONLY` field_behavior.

Reviewer 2: ```suggestion
Server owned fields **should** be indicated with the `OUTPUT_ONLY` field_behavior.
```
I'm guessing this is guidance.

Author: I modified all of these to **must** because **should** doesn't actually mean much

> All other types of fields should be considered to be owned by the client. The

Reviewer 2: ```suggestion
All other types of fields **should** be considered to be owned by the client. The
```
I'm guessing this is guidance. If not, maybe change the word to `can` as in `can be considered...`

> server should respect the value (or lack thereof) for all client owned fields

Reviewer 2: ```suggestion
server **should** respect the value (or lack thereof) for all client owned fields
```

and not modify them. Failure to respect client set values can lead to

> [Declarative clients][] entering an infinite loop to correct the change.

Reviewer 2: I think the last sentence belongs in the Rationale section.

Author: Added Rationale section for Single Owner Fields

### Effective Values

There are instances where a service will allocate, generate, or calculate a
value if the client chooses not to specify one. For example: a client may

> create a virtual machine without specifying a static IP address for the virtual

Reviewer 2: ```suggestion
value if the client chooses not to specify one. For example: a client
creates a virtual machine without specifying a static IP address for the virtual
```

Just removing the "may" to make it less ambiguous and avoid it being interpreted as guidance.

> machine to be available on. Such a scenario is opting into dynamic IP address

Reviewer 3: There seems an important and subtle distinction between this scenario (in which the system ends up managing the address), and a related scenario in which the system assigns an IP address on behalf of the client, but in which the IP address remains fixed (i.e., has the same semantic as if the caller had selected that address).

Author: Those are different situations, but in my mind are handled in the same way. By splitting the client input and the server value for a field that can be dynamic or statically assigned by the server if not specified by the client I think both problems are solved

allocation.

Some examples of these types of fields are ones that are:

* generated (UUID)
* allocated (dynamic IP address)
* assigned (most recent software package version)

An attribute with an effective value ***must*** be expressed as two fields in
the API:
* a mutable field that can be optionally set by the user and **must not** be
modified by the service
* an `OUTPUT_ONLY` field that records the effective value decided on by the
service

Example:
```proto
message VirtualMachine {
  …
  string ip_address = 4;
  string effective_ip_address = 5 [
    (google.api.field_behavior) = OUTPUT_ONLY
  ];
}
```

#### Naming

Effective values **must** be named by prefixing `effective_` to the mutable
field's name.

### User-Specified Fields

For user-specified fields, the value in response from the service **must** be
the same as provided by the create or update request. For string fields this
means returning the value unchanged, with two exceptions: 

* When a string field has a data type annotation, a normalized string that

> represents the given value may be returned.

Reviewer 2: ```suggestion
  represents the given value **may** be returned.
```

> * When a string field has a default value indicated by the `default` keyword,

Reviewer: I think an example would help here, currently noting that it's proto2 only.

Reviewer 2: Since Google APIs must be `proto3` I'm not sure this is relevant. Are there services Terraform supports that have published proto2 schemas?

> * When a string field has a default value indicated by the `default` keyword,

Reviewer 3: Apologies if this comment is off-base, as I'm not familiar with the constraints of protobufs, but per my comment above regarding dynamic vs. defaulted-static IP address allocation, there are scenarios where the value has a true default but that value is not fixed.  Phrased differently: there seems an important distinction between a server-calculated default for a property that remains client-managed, and a default that causes the property to become system-managed (such as a DHCP-managed IP address).

Author: Is that distinction important?

From my side it's important to determine what fields can be set by a state-driven client and what fields we expect the server to communicate information to the client via server-set fields. Those should not mix. Both a server-calculated default IP address *or* a dynamic IP address can be examples of server-set values that can be communicated via output-only fields

Reviewer 3: To my view, the key distinction is whether the property is _managed_ by the client or by the server.  A property  that has its value set at resource creation but will never be subsequently changed by the server remains client-managed.  This is the case for the exception carved out above for `default`, but is also true for more complex scenarios that cannot be expressed through a `default` clause but nonetheless represent a one-time initialization of a property that remains client-managed.

> that default value may be returned when the given value is empty or missing.

Reviewer 2: ```suggestion
  that default value **may** be returned when the given value is empty or missing.
```

### Normalizations

Only a set of allowed normalizations may be applied. Values that are not annoted
with the `google.api.value_type` annotation **must not** be normalized in any
way.

#### Allowed Normalizations

This list of allowed normalizations is defined as follows:

* uuid
* ipv4
* ipv6
* email

> Normalizations on fields **must** be described using the `google.api.value_type`

Reviewer: I think this is the only contentious point left in this PR: the rest looks good.

We can't really introduce guidance without an annotation that goes with it. We could introduce the annotation easily enough: should we just do that? 

This will of course mean that clients will not support these normalizations, at least for a while. But that's probably ok: declarative clients could probably add support for these first, with other normalizations to follow as needed?

Reviewer 2: Let's start drafting an annotation! My two cents would be:
* let's make it of type `message` so that we can extend it in the future if necessary
* let's use an enum for the supported formats

```
message FieldValue {
  enum Format {
    FORMAT_UNSPECIFIED = 0;

    IPV4 = 1;
    IPV6 = 2;
    EMAIL = 3;
    UUID4 = 4;
  }

  Format format = 1;
}

extend proto2.FieldOptions {
  // Example:
  // (google.api.field_value).format = IPV4
  FieldValue field_value = xxxxx;
}
```

Reviewer 2: After the API Design team discussed this, I volunteered to draft an annotation, [take a look](http://cl/557929054) (fyi internal link)

Reviewer: I think once that goes in, let's update this PR with some examples.

Reviewer 2: The annotation is now submitted internally, but isn't published yet. This AIP can be updated to mention it, but it should be left in `draft` until we have "launched" the annotation.

> Only a set of allowed normalizations may be applied. Values that are not annoted
> with the `google.api.value_type` annotation **must not** be normalized in any
> way.
> 
> #### Allowed Normalizations
> 
> This list of allowed normalizations is defined as follows:
> 
> * uuid
> * ipv4
> * ipv6
> * email
> 
> Normalizations on fields **must** be described using the `google.api.value_type`
> annotation.

Reviewer 2: ```suggestion
A field that is normalized by the service **must** be annotated with the
`google.api.value_type` extension. The allowed set of normalizations includes
the following formats:

* uuid
* ipv4
* ipv6
* email
```

This is a suggestion to tighten up the guidance.

Author: Done

> #### Field Value Handling

Reviewer: isn't this redundant with the `user-specified field` handling above?

I'd just remove or consolidate it.

for differences.

For example, in a resource that accepts an email address on a particular field
a client may specify a given email address in a variety of ways. For the email
`ada@example.com` a client may choose to specify `ADA@example.com`,
`aDa@example.com`, or `AdA@example.com`. These are semantically identical and
*should* all be accepted by the service. The service then may choose to
normalize the email address for storage and retrieval through downcasing or
canonicalization. Importantly, the information surfaced to clients on the
normalization of a field will not describe the normalization algorithm itself,
but instead the comparison method used to accurately compute if two values
should be considered equal.

### Rationale for Field Value Handling

For fields not using an allowed normalization,
[Declarative clients][] will not be able to identify which changes are
semantically meaningful. When a [Declarative client][Declarative clients]
sends a particular value it will ensure that the value is being returned by the
service to validate it was set correctly.

[Declarative clients]: ./0009#declarative-clients

## Changelog

> - 

Reviewer: I'd remove this: there's no entry here, so markdown may format it strangely.

---
id: 129
state: draft
created: 2023-06-21
placement:
  category: resource-design
  order: 65
---

# Server-modified Values and Defaults

### State-driven clients (aka Convergent clients)

> Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.

Reviewer: can this be moved to https://google.aip.dev/9 instead? It feels a bit weird to define a new term in an AIP that's focused on server modification of values.

Also consider merging terminology. We use the term "IaC" for these type of clients, but perhaps it's best to refactor them all to "state-driven".

Reviewer 4: Adding this to the glossary is a good idea.

> ### Permadiff

Reviewer: add to AIP-9? Despite being a declarative term, I think the concept broadly applies when describing this behavior to APIs overall.

Reviewer 5: Yes, and IMO (from comment above) the broad concept is a "single owner" principle for any information. The term "permadiff" might not be universal, but it is a clear example of the oscillation that happens when you violate the simple owner principle.

Author: I added single owner fields to AIP-9.

Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.

## Guidance

### Server Side Defaults
There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.

#### Dynamic Defaults
When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
* generated (UUID)
* allocated (dynamic IP address)
* assigned (most recent software package version)

> An API service ***must*** treat a client-side dynamic default field as read-only and must provide a separate `OUTPUT_ONLY` field that reflects the server assigned, allocated, or generated default value. 

Reviewer 6: Should "read-only" be "write-only" here, or am I misunderstanding something? _Ideally_ they'd be read-write but never canonicalized, I think, and could potentially represent a constraint, but write-only works.

---

When I say a constraint, I mean something like expressing a CIDR in the example below, and having the service allocate an ip in that range. For example:

input:

```
IpAddress = "192.168.1.0/24"
```

output:

```
IpAddress = "192.168.1.0/24"
ProvisionedIpAddress = "192.168.1.7"
```

Reviewer 4: The text was intending to say the service shouldn't populate empty input fields or change values provided by the client.

Reviewer 6: Oh, got it, read-only from an implementation of a service perspective and not from the frontend of the API perspective. That makes more sense, thanks!

I think this could use a little more front matter to explain the approach. Something along the lines of "An attribute with a dynamic default **must** be expressed with two fields in the API, a mutable field that can optionally be set by the user, and an output-only field that records the final value decided on by the API service."

Author: Added wording specifying two separate fields

Reviewer 6: Works for me, thanks!

Example:
```proto
message VirtualMachine {
  …
  string ip_address = 4;
  string provisioned_ip_address = 5 [
    (google.api.field_behavior) = OUTPUT_ONLY
  ];
}
```

#### Static Defaults

> Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version (#TODO slevenick: link to AIP 180 once merged). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default in effect creates a new type with a value designated as an equivalent to an empty value.

Reviewer 4: In the meeting today, "creates a new type with a value designated as an equivalent to an empty value" was not sufficiently clear to everyone what it meant. The default value is not populated in the field. This is effectively machine-readable documentation, similar to field behavior annotations.

Author: Changed wording here, see if this makes more sense

Reviewer 4: Works for me. Thanks.

---
id: 129
state: draft
created: 2023-06-21
placement:
  category: resource-design
  order: 65
---

> # Server-modified Values and Defaults

Reviewer 5: Thanks for writing this!

I think it would be fantastic to frame fields as having a "single owner" principle, before going into client-server and permadiff as an example of problems from violating that principle.

The principle is that each field of a resource may be determined by one-and-only-one software component. The obvious problem being when any field is owned by two components the resource will oscillate between states.

Because GCP is a client-server architecture, being output_only has a clear meaning. And for cases where there's a single client the most apparent oscillation is between client and server. 

But in other cases the principle stands up --- in k8s controllers are also clients, so status fields are writable, but only one component-client is properly entitled to modify any given field. (field manager metadata helps track that.)

And even in client-server cases the principle extends to multiple clients. If there are any read-write fields which a non-Terraform software component will be driving, then the Terraform client must be told explicitly to ignore_changes for converging state.

Author: I did a large rework and used the single owner framing

> ### State-driven clients (aka Convergent clients)

Reviewer 7: I think it would be good to have a brief paragraph of text directly under the top-level heading, and these two paragraphs *might* be better under a rationale. I'll leave that sort of style call to @toumorokoshi though.

Reviewer: agreed there should be some single-paragraph forward.

Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.

### Permadiff
Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.

## Guidance

> ### Server Side Defaults

Reviewer 5: Can we break away from the terms "Server side default" and "Dynamic default"? They really both water down the meaning of a field's default value.

That is to say - a default value is a fact, not a change in the data.

If you have a timeout in seconds, and the default is 60, then in that field the integer 60 --- and any form of zero, nil, or empty --- are semantically indistinguishable.

Given that - the "server side default" is the effect where marshalling either empty and 60 the same value. The server should as a best practice prefer to marshal as the integer 60 in responses (esp for the benefit of weakly-typed clients that don't have access to field default metadata).

Either the server or any client of that field (or customers in their own files) may freely choose between 60 and leaving empty or missing, and no combination of those values would compare as "different" from each other. (e.g. adding or removing "timeout: 60" to a file would not cause a "no-op update" in the next apply)

Reviewer 6: We'd need a replacement term, which would be fine if it's clear. You're defining a default as a static/transport (i.e. protobuf) default. That's reasonable- Terraform assumes the same is true, for example- but doesn't necessarily match up to what user-facing documentation treats as a default

Classifying dynamic/server-side/complex defaults as a type of default makes sense given a practical example imo. For example, a field with a description like:

> The version of Kubernetes running on this NodePool's nodes. If unspecified, it defaults as described [here](https://cloud.google.com/kubernetes-engine/versioning#specifying_node_version).

Author: As Riley points out there is a need for a more complex type of default that cannot be captured as a static value. That's what is meant by "dynamic default", where the value that is assigned by the server when the field is unset changes over time

Reviewer 5: Yep, understood.

The GKE example you had seems to have a good example - the `initialVersion` which may be given values like `latest` and `1.X` . Those aren't defaults - you wouldn't even call them missing-or-empty zero values. But what is ideal is if the output_only `currentMasterVersion` contains the effective/calculated value - and the `initialValue` retained the configured `latest` or `1.X`. That's a great example of the two-field approach where the client-side and server-side aren't fighting over field ownership.

You may also go a step further and say that the default value of `initialVersion` is `latest`. That's where the string `latest` has the same meaning as the missing-or-empty zero values.

The problem has been historically that field value hasn't been recognized as a valid configuration - and the server-side has been overwriting the client's value with the result of an algorithm which may change over time. Saying the default value is `latest` is very different from saying there's a dynamic default that the service will overwrite your field with.

Re: static/transport (protobuf) default --- that's also slightly different. What protobuf calls a default is the value that's elided when marshaling and introduced when missing. That's also a bit destructive to data, because a payload can't round-trip without having default values become absent...

I'd be suggesting more that the openapi definition of default --- which is based on knowing what value is semantically equivalent to missing-or-empty zero values.

Author: Changed

> There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.

Reviewer 7: I don't personally see the connection between caching and a software version. I'd be happy to try to think of another example if that would help. (Number of replicas for a distributed system might work, for example.)

Reviewer 4: I think this is about hosted open source, such as Redis release 7.0.11.
https://redis.io/download/

Reviewer 7: I suggest that's either clarified, or a different example is picked - readers shouldn't have to be joining dots.

Reviewer 4: GCE has lots of examples, such as the default machine type.

Reviewer 5: It's also a bit of a classic anti-example. Exactly the case where you should use two fields instead of one. 

Better would be an example where the default must not evolve for the lifetime of the field. Like a timeout, or a non-enum select-string like Sku: "Basic"

Author: I'm simplifying this to what I think is a very clear case of server-side value assignment: a GCE VM's IP address.

Either the address is set by the user or chosen by the server. The server-assigned value cannot be static and the final field must have a value for the VM to be usable.

> #### Dynamic Defaults

Reviewer 5: I'd recommend calling these "Effective values" rather than "Dynamic Defaults".

As a metaphor, if we had prior art involving "Dynamic Constants" and "Static Constants", that would be misleading. A "static constant" is just a "constant", and a "dynamic constant" is just a "variable", and describing those situations as two kinds of constants is misleading/counterproductive.

In other words - there is no such thing as a "dynamic default". But, like you describe, there may be:
* a normal field and an effective value pair
* the normal field may allow a setting which means the server allocates or chooses the effective value
* the default value for the normal field may in fact be that setting where the effective value is expected to be allocated or algorithmic.

Author: I think that "dynamic default" describes the way these types of fields are handled today, where the default is determined to be a non-static value.

I'll change to effective values though, as the proposal is to split them into a normal field + output-only effective value

Reviewer 5: Much appreciated! That's the only reason I'd feel strongly about terminology at the end of the day, 

When a client configured "", what that value "" meant is "dynamic" or "calculated" or "latest", and the service would overwrite the fact of that configuration with an effective value that resulted from an algorithm or allocation... Essentially a lossy change, causes a permadiff, but throws out the fact that the cluster was configured for "latest" rather than what actually took place.

Then we can reserve the term "dynamic default" for that existing anti-pattern, because that's how it's used today, and not need to clarify the difference between "old dynamic defaults" and "new dynamic defaults".

Reviewer 7: We could always clarify it with "dynamic defaults expressed as effective fields" to avoid ambiguity.

Author: Changed to effective values in all places

> When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 

Reviewer: nit: space after the heading.

* generated (UUID)
* allocated (dynamic IP address)

> * assigned (most recent software package version)

Reviewer 5: Would recommend:
```
Fields with an effective value must be expressed as separate fields in the API:
* mutable field(s) configured by the user
* output only field(s) that reveal the effective value(s) which resulted
```

for specific cases

```
string ip_address;
string effective_ip_address [(google.api.field_behavior) = OUTPUT_ONLY];
```
In this case you would say the ip_address empty-value "" is defined meaning dynamically allocated. You would also say the default value of ip_address is the "" empty-value.

An API designer even plan in advance and say that the ip_address string "dynamic" means dynamically allocated, and that the default is "dynamic". 

```
string ip_address [(google.api.default) = "dynamic"];
string effective_ip_address [(google.api.field_behavior) = OUTPUT_ONLY];
->
POST { ... provide no opinion ...}
GET { "ip_address": "dynamic", "effective_ip_address": "10.5.29.234" }
```

That would give you even more precise resource response output, and enable different kinds of non-literal IP allocation, e.g. ip_address = "dynamic:10.5.0.0/16"

a real-world example of that would be in the hosted-software-version field where the default value of the setting means "server's choice", but there's also "server's choice of cutting edge version" (release channel).

```
string minecraft_version [(google.api.default) = "stable"];
string effective_minecraft_version [(google.api.field_behavior) = OUTPUT_ONLY];
---
POST { ... provide no opinion ... }
GET { "minecraft_version": "stable", "effective_minecraft_version": "3.45" }
---
POST { "minecraft_version": "2.67" }
GET { "minecraft_version": "2.67", "effective_minecraft_version": "2.67" }
---
POST { "minecraft_version": "unstable" }
GET { "minecraft_version": "unstable", "effective_minecraft_version": "3.89-preview.2" }
```

Author: Modified. Do we want to show an example of how to use an automatic updating version tag? I think it's unclear how a "version": "latest" will resolve as the latest version changes over time. Does the server automatically upgrade to keep you on the latest, or is it a one-time thing? We need some clear guidance here to avoid contrasting implementations

An attribute with a dynamic default ***must*** be expressed as two fields in the API:
* a mutable field that can be optionally set by the user

> * an immutable `OUTPUT_ONLY` field that records the final value decided on by the API service

Reviewer 7: By "final" do you mean that this can't change later in the resource's lifecycle? I can imagine a couple of reasons why it might change:

- The client performs an update, explicitly specifying a value (and then perhaps another which clears the field again, going back to "server-decides")
- The dynamic value could change - for example, the client-specified field being unset could mean "the current latest software version, at any point in time" in which case a client fetching the resource twice without modifying it between fetches could see the field value changing.

Reviewer 4: "Final" means that the service won't change it after choosing the value -- that's what makes it a default value. But we could probably remove that term.

If the value can be changed by the service, that should just be reported in a regular output-only field.

Reviewer: not to bikeshed, but I agree that we should just remove the term "final" - I don't think it adds much to the meaning of the sentence.

Reviewer 7: @toumorokoshi: I disagree - if it's intended to be read as "this won't change" then it's really important. But I'm not sure whether that actually meets the use cases.

Consider the "potentially-dynamically-assigned IP address" use case:

- The user can specify an IP address if they want
- If not, the server will allocate an IP address
- After some state changes (e.g. it's a VM which is rebooted, or a DHCP server and the lease has expired) a different IP address is allocated

I *thought* that the intention was for this pattern to be used for that use-case... but it fails the "service won't change it after choosing the value" criterion.

Let's work out the desired semantics, then wordsmith it.

Reviewer: I agree let's figure out the behavior and take it from there.

I think that you're describing a behavior that falls in scope of server-modified values, but should be it's own line vs mentioned  on the side from the output-only field.

We're saying two things then:

1. you must have an `OUTPUT_ONLY` field that pairs with the input field.
2. this output_only field cannot change it's value over time. 

Both have different reasons for why they're designed that way. I'm not 100% sold on 2, but can certainly agree to 1.

Reviewer 4: A dynamic IP address that can change on reboot wouldn't be a default value of an input field. That should just be an output-only field.

Reviewer 7: But aren't we *talking* about an output-only field here? (The second field in the list, where the first is a mutable field.)

Reviewer 4: We're talking about an input+output scenario, not just output. Without the option for optional input, it wouldn't be a dynamic default.

Example:
```proto
message VirtualMachine {
  …
  string ip_address = 4;

>   string provisioned_ip_address = 5 [

Reviewer 7: Do we want to give any guidance on the naming of the fields?

Reviewer 4: Good idea.

I would recommend "effective_" as a prefix for the output-only field, in general. There are some common cases of dynamic defaults for which we could recommend more specific alternatives, such as "provisioned_", "allocated_", or "assigned_" for assigned values.

Reviewer 7: I like `effective_` too - that's certainly a convention we happen to use in the .NET client libraries (with nothing to do with protos normally) for "I've got inputs from a number of places, potentially including defaults, but this is what I'm really using".

Reviewer: could that be a new section in this AIP? I'm fine with recommendations on naming as long as we don't make it a hard requirement. Or if it is a hard requirement, then we choose a single prefix for consistencies sake.

Also I would actually argue that it should be a suffix: this ensures that the input value and the output value are side-by-side when sorting them:

```
ip_address
ip_address_provisioned
```

Reviewer 5: +1 for effective --- good for meaning "the value in effect" regardless of whether it was allocated, applied, generated, selected, or otherwise determined.

Re suffix - I do appreciate the sorting concern - though adjective-first may have precedence. e.g. create_time/update_time vs time_created/time_updated.

Author: Added a prefix naming suggestion as a sub-section on effective values. Happy to move that around if there's a better structure to be had

    (google.api.field_behavior) = OUTPUT_ONLY
  ];
}
```

#### Static Defaults

> Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.

Reviewer 7: Nit: put google.api.default in backticks?

This is a new annotation that doesn't exist yet, right? Do we need a different annotation for each field type? Can it handle messages?

Author: Yes, this is a new annotation. I'm not positive on how this will be implemented yet, first trying to get agreement on what *should* be implemented

> Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.

Reviewer: ```suggestion
A service ***may*** choose to designate a value to be equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.
```

also this paragraph should be truncated to lines of 80 chars each.

Reviewer: note we can't merge this in until a `google.api.default` annotation exists. So that work should be done first.

Reviewer 5: +1 the missing `google.api.default` leaves a large gap in the interface description.

> Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.

Reviewer: could we have a decision tree (it doesn't have to be a diagram, written is fine) about when an API author should opt to use a static default vs a dynamic default?

Putting the guidance of when to use the value in the sub-section forces the reader to read the whole document before knowing which section is relevant.

Example:
```proto
message VM {
 string sku = 1 [(google.api.default) = "foo"];
}
```

> ### Mutable fields

Reviewer: should we call this `user-specified fields` to align on the terminology in AIP-133? https://google.aip.dev/133#user-specified-ids.

I think fewer synonyms are great. also "user-specified" is more precise: mutable fields may imply it's mutable by the server as well.

Author: Changed to "User-specified"

> API Services ***must not*** modify mutable fields.

Reviewer 7: This feels a little terse - I'm not sure whether I'm missing something. Presumably we've seen examples of where this *has* been done in the past, in which case it would be good to give more of an idea of why a service might *want* to do this, and why they shouldn't.

Reviewer 4: You're asking for a Rationale section, which is fair.

Author: Added some rationale in the first paragraph of this AIP along with a general rework. Does that work?

> ### Semantic Strings Handling

Reviewer 7: I'm not keen on the term "semantic string" here - it's at least not one I've come across before. This sounds like it's really talking about a canonical or normalized representation, but I may be missing something.

Reviewer 4: "Semantic string" was the term used to describe a data type represented in the API using a string: uuids, IPv4 addresses, DNS RFC1123 names, etc. These have specific value validation rules and sometimes are represented in systems in ways that don't preserve the original client-provided string values.

Reviewer 7: Right. I think it's fine as a concept - I'd just prefer to use well-known terms... or if we're *going* to define a new term, make it really clear that we're doing so. (A reader shouldn't read the section and feel bad for not knowing what a term means already, if we've just made it up...)

Reviewer 4: I don't love the term. I was just explaining it. :-)

I'd probably just use something like "data types represented as canonicalized strings".

Reviewer 5: "Field Value Handling"? "Semantic Value vs Marshaled Representation"?

It applies to strings, but isn't limited to them. Not certain what the term should be but should definitely cover all of those cases.

Personally, I think of this as "Value entropy vs Marshaled entropy", which isn't more clear.

But with a uuid for example, the field value has 2^128 distinct values, exactly the same value space as 32 hex characters (16^32). Literal comparison of [0-9a-zA-Z] has more distinct values (22^32) which is why you *must* perform all equality+comparison operations on the value space, not in the representation space.

This is really good to connect to the above concept also. Having a defined default on a field actually means that it _is not a server modified value_ --- they are different representations of the same value.

```
enum ColorEnum {UNSPECIFIED, RED, GREEN, BLUE};
ColorEnum color; 
```
The entropy of color is 4, because the field may be in the states of `{UNSPECIFIED, RED, GREEN, BLUE}`

Or a different angle with string rather than enum mechanically:
```
// optional field, valid colors include: "red", "green", "blue"
string color_name; 
```
The entropy of color is 4, because the field may be in the states of `{empty, "red", "green", "blue"}`. Even if you tolerates "Red" and "RED" the entropy of the value space is still 4.

It's when you declare the default value - you've defined the semantic value of the empty representation.
```
enum ColorEnum {UNSPECIFIED, RED, GREEN, BLUE};
ColorEnum color [(google.api.default) = ColorEnum.RED]; 

// valid colors include: "red", "green", "blue". default is "red".
string color_name [(google.api.default) = "red"); 
```
The entropy is now 3, because (UNSPECIFIED==RED) and (""=="red") _for all intents and purposes_.

So if you send number's json text of 7.000 and get back 7, uuid with "1234abcd" and get back "1234ABCD", send color of `UNSPECIFIED` and get back `RED`, or `""` and get back `"red"` ---- the key concept is that _the server has not modified the value_.

Author: Numbers are an interesting case of this, we sometimes see odd behavior in service's decisions about precision and trailing zeros: https://github.com/hashicorp/terraform-provider-google/issues/15166

I'm not sure how to handle for this. Potentially a defined type for duration and we can implement the comparison client-side?

Overall, I changed this section to not be specific to strings as numbers have the same issues

> For a number of string transmitted value types an API service may be required to convert strings into other formats. For example, a UUID value may be transmitted from the client as an uppercased string `"3BBAB074-4E18-41F1-AD21-B7358B3DB0CE"`, stored in a database as a sequence of bytes and transmitted back to the client as a lowercase string `"3bbab074-4e18-41f1-ad21-b7358b3db0ce"`. The semantic meaning of such a string doesn't change however state-driven clients have no way of determining such differences. To avoid such mismatching values:

Reviewer: please break this into 80 char lines.

Author: Done

* service must define and validate a single representation of a semantic string. 

> * service must return an error when a client submitted value doesn't match defined format.

Reviewer 7: This feels like it hurts imperative clients for the sake of declarative clients. I can absolutely see why declarative clients would want this, but I'd prefer not to put a burden on other clients.

I don't remember seeing a discussion on this requirement before - which doesn't mean it hasn't happened, of course. If you could point me (internally) to any docs around this, that would be helpful.

Reviewer 4: This isn't just about declarative clients. When the service mutates values, that makes it difficult to determine equivalence and correctness without detailed, hardcoded knowledge of the service, which may even need to be reverse-engineered through trial and error.

It affects any clients that use the state in a general way to make decisions or to compare with a "golden" representation. This includes declarative tools, but also policy engines (e.g., Open Policy Agent), security response agents, controllers, data-driven test frameworks, recommendation tools, LLMs, and so on. 

It even affects "imperative" clients that aren't manually coded. Let's say a GUI has a form to update a particular value. The user enters "x", which is allowed, and the service changes it to "X". Did the update succeed? It depends on how the GUI determines completion/success and whether a success response was recorded. 

Input/output values are not a well supported concept in many tools outside the Kubernetes ecosystem.

Reviewer 4: The easiest thing to do in many cases is to just make validation more strict -- disallow values that would be transformed.

Reviewer 7: That's easiest for *some* clients, but definitely not for all.

If I were writing a GUI performing an update, I'd expect that an update RPC which succeeds has performed the requested update. If I want to find the canonical representation, I can look at the response. I don't need to include that canonicalization code in my application.

With the restriction here, *every* application has to perform canonicalization unless their "natural" input happens to be the canonical form. I strongly expect the correctness of that canonicalization code to be a hit-and-miss affair. Even the fact that canonicalization is required is likely not to be obvious to any but the most diligent app developers until they first receive an error.

There's a *potential* mitigation for client library users if we have a fixed set of canonicalized string formats: we can provide canonicalization code once per language. (We could even *potentially* apply canonicalization transparently in the client, but that feels a bit too much like "magic" for me.)

It's possible that strict validation is the least-worst option, but I think it merits further discussion - and I'd like to acknowledge the trade-offs within the rationale.

Reviewer: I think +1 for some set of defined data types. We could start there, I'm sure that catches 90% of the situations where value normalization occurs.

The other 10% would be bespoke normalization, and validation is a good fallback. 

I don't know how opinionated we have to be here really. handling normalization or defaulting of user-specified values can be broken down into several options, of which the service teams can choose one:

1. choose a well-known data type, enabling some normalization or validation client-side. Don't normalize on the server, validated and reject.
1. don't normalize values anywhere, instead validate and reject malformed ones.
1. normalize or default values, and provide a paired OUTPUT_ONLY value to see the final result.
1. choose a specific default value, document it, and never change it.

Reviewer 4: Clients can be categorized on multiple axes: special case vs generic, manually constructed vs automated, static vs dynamic, management vs data plane, operation-driven vs state-driven, connected vs at rest, and so on.

The case you're describing sounds like a special-case, manually constructed, static, connected, operation-driven management client. To understand the canonicalization for that specific field of that resource type of that service, a human would have had to read documentation or proto comments, analyze gcloud source code, and/or discover the behavior through trial and error. 

Clients that are passing data values from one system to another generally would have to transform and reformat the data. That's normal.

Clients that are taking input values from users could just request values within the valid range. That's also normal.

Services transforming resource attributes provided by the client violates the principle of least surprise as well as obstructing generic, automated, state-driven clients. The behavior is different from typical fields, which are not transformed by services, and currently is mechanically undiscoverable.

We could standardize a "strict" parameter on Create and Update to require pre-canonicalized values, but I'm unconvinced that that complexity is necessary.

Another approach could be to handle them the same way as dynamic defaults, with an "effective_" output-only field.

Reviewer 5: Yeah, this would be a case where there's a "Value Type" is of a more specific class than the in-transit/in-memory field type holding it. Especially because of cases where strength of typing in-transit on different protocols already varies - ability to GET/PUT json representation means map/array/string/number is always the most extreme (and universally supported) case. 

I like the approach of thinking in terms of normalization and validation first.

**Validation** is really an invariant. The client doesn't need to validate, but the service will of course only accept a valid resource reference for a ref string field. I think the robustness principle would apply there "be conservative in what you do, be liberal in what you accept from others" --- so accepting upper-case/lower-case UUID, etc, would be better than requiring the client to mirror exact marshaling of fields in all cases.

It is true you could say there's one way to solve "value-entropy vs representation-entropy" by saying there's one single acceptable representation for each valid value... But that would put callers in a position where they could be fighting the format of their underlying inputs... e.g. to_lower(guidgen())... So I'd suggest letting the representation entropy be larger and accept more than one well-formed representation of the same value.

**Normalization** is where it gets interesting. With (un)marshalling going from representation-space to value-space and back to representation-space ---- the normalization that happens may not even be intentional. The only way to avoid any normalization is to accept the highest fidelity representation all the way from client to backend, and validate incoming but only normalize server-side at a point past what will ever ever round-trip. e.g. store strings and only convery to uuid/timestamp in what's passed to control plane.

Which is a lot... And also re "...be conservative in what you do..." you might expect that the service's job _is_ to have an extremely consistent set of representations... e.g. if you are writing client code to do evaluate policy over resources - it would be a kindness to have a field in the same representation every time rather than preserving the differences in representation that are not differences in value...

**Equivalence** is the last bit - where normalization causes problems is in value-equality, aka representation-equivalence... For a client which must compare two values for equivalence, what operator must it use?

The advantage of the service opting to heavy normalization in responses is it help --- when a client has two responses from the service --- it's more likely it can use naive string equality on fields to compare them...

In the case of a convergent client that's reconciling or doing drift detection --- that's really the case where it needs to know how do do value-space equivalence.  It's comparing external (input) values to returned (normalized) values...

Which means, rather than string==string, it must know "what equality operator should values of this field be compared using?" --- Nice thing is by taking into account a dozen-ish "kind", "default" --- that should knock out vast classes of permadiff.

Ah, also --- unordered list --- that would also be a field "kind"-ish annotation that's taken into account by client value-equivalence operator.

TL;DR suggestion...

* services should accept the broader set of representations of a field value in requests
* services should produce the same representation for the a field value in responses
* annotation must indicate when a string field contains a more specific kinds of value
* converging/annealing clients must implement comparison operators for kind/default annotations
* aip needs to define about a dozen a-priori kinds

(IMO dynamic kinds, regex value extraction, all out of scope. complexity explodes, diminishing returns. etc.)

See also... [7.3](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00#section-7.3).  Defined Formats https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00#section-7.3
 For a starting point of "kind" annotations... You'd also consider a "resource_reference" to also be an example of where we've already added a "kind" marker onto a string field.

Author: Added a set of defined data types for strings
